### PR TITLE
VecDeque: drop remaining items on destructor panic

### LINKED ISSUE #67232: `VecDeque` leaks remaining content when one destructor panics
STR:

```rust
use std::collections::VecDeque;

struct D(u8, bool);

impl Drop for D {
    fn drop(&mut self) {
        println!("Drop {}", self.0);
        if self.1 {
            panic!("panic in `drop`");
        }
    }
}

fn main() {
    let mut q = VecDeque::new();
    q.push_back(D(0, false));
    q.push_back(D(1, false));
    q.push_back(D(2, false));
    q.push_back(D(3, false));
    q.push_back(D(4, false));
    q.push_front(D(100, false));
    q.push_front(D(101, false));
    q.push_front(D(102, true));
}
```

Output:
```
Drop 102
thread 'main' panicked at 'panic in `drop`', test.rs:9:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
Drop 101
Drop 100
```

Caused by the `Drop` impl of `VecDeque`:

https://github.com/rust-lang/rust/blob/90b957a17c1abba979aa41234ce0993a61030e67/src/liballoc/collections/vec_deque.rs#L146-L154

Apparently `drop_in_place` will continue dropping elements when one destructor panics. That doesn't carry over to the second call, of course.

### PR DESCRIPTION
**Pull request by jonas-schievink**
Closes https://github.com/rust-lang/rust/issues/67232

### CODE REVIEW

File: src/liballoc/collections/vec_deque.rs [hunk 1]
```diff
@@ -144,11 +144,21 @@ impl<T: Clone> Clone for VecDeque<T> {
 #[stable(feature = "rust1", since = "1.0.0")]
 unsafe impl<#[may_dangle] T> Drop for VecDeque<T> {
     fn drop(&mut self) {
+        struct Dropper<'a, T>(&'a mut [T]);
```
Comment (KodrAus) [inline src/liballoc/collections/vec_deque.rs]: Let's leave a comment here that discusses what this `Dropper` is for.

Comment (jonas-schievink) [reply src/liballoc/collections/vec_deque.rs]: Done. I hope this makes it more clear.

**Pull request merged**